rfcs/032-calm-deletions/README.md

```diff
@@ -20,14 +20,26 @@ The deletion watcher will consist of:
 
 The lambda will place a message on the reindexer queue requesting a "reindex" of Calm records to be sent to the worker. It will be triggered by a scheduled Cloudwatch event or manually. By default it will request the full source scan but in the case of a manual trigger it will be possible to request specific IDs, which will be useful in the case of expediting known deletions.
 
-The worker will consume records, filter out those that are already flagged as deleted, and check for the existence of the rest in the Calm API. If the record has been deleted (is no longer present) then it will update the VHS entry to flag the deletion. The deletion flag will live in the DynamoDB record so that the flag can be checked without fetching the object from S3.
+The worker will consume records, filter out those that are already flagged as deleted, and check for the existence of the rest in the Calm API. If the record has been deleted (is no longer present) then it will update the VHS entry to flag the deletion. The deletion flag will live in the DynamoDB record so that the flag can be checked without fetching the object from S3. After updating a record, the worker will notify the transformer topic.
 
 The calm transformer will check for the presence of the deletion flag in the source data and create `Deleted` works as appropriate.
 
 This architecture is illustrated below (arrows indicate the direction of data flow):
 
 ![image](https://user-images.githubusercontent.com/4429247/106171539-8da8dd00-6189-11eb-9aca-2577f1ab6ed7.png)
 
+### Minimising Calm API queries
+
+Checking for the existence of the records can be done naïvely by performing a search (the only relevant action available to us in the current API version) for a given record ID and confirming that the returned `SearchResult` number is equal to 1. Obviously, this requires 1 request per record.
+
+We can, however, do better than this. Consider that we have a large population of records with a fairly low prevalence of deletions. In one query, we can search for a set of `N` records, and know that the difference between `N` and the number of results is the number of deleted records in that set. If there is no difference, we can move on immediately.
+
+At this point we could iterate through either the results or the record IDs to find which are missing, or we could find the missing records via binary search - but the former means that we only reduce the number of queries if there are _no_ records missing among the set, and the latter requires both that the resultant queries are made sequentially (not in parallel), and more significantly binary search is very poorly suited for finding multiple occurences.
```

Comment:
typo:

```suggestion
At this point we could iterate through either the results or the record IDs to find which are missing, or we could find the missing records via binary search - but the former means that we only reduce the number of queries if there are _no_ records missing among the set, and the latter requires both that the resultant queries are made sequentially (not in parallel), and more significantly binary search is very poorly suited for finding multiple occurrences.
```

more significantly:

> binary search is very poorly suited for finding multiple occurrences

I can imagine this is true if records are distributed uniformly throughout the population; I wonder if it applies to this particular problem? My guess is that records probably aren't being insta-deleted at a uniform rate throughout Calm – more likely we'll see particular runs of records that were all created at the same time being deleted in one go.

Reply:
There is nothing to say that we'll receive items in the order that they were added - a DynamoDB scan is unordered (not sure if it is usually predictable in actual use though).  

Binary search is poorly suited for finding multiple occurrences because it's a search, it finds single items - nothing to do with the distribution. The linked review paper covers probability distributions and different priors as well as different regimes of sparsity if you're interested; I've actually realised that because we know the number of deletions for _every_ query there are better algorithms than the given one and will be updating this to reflect that.

Reply:
> There is nothing to say that we'll receive items in the order that they were added - a DynamoDB scan is unordered (not sure if it is usually predictable in actual use though).

Yeah, I should have clarified that this is a theoretical curiosity; we have no good way to order Calm records by modified date.